disorder-fsm/src/Disorder/FSM.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE LambdaCase #-}
+module Disorder.FSM (
+    Transition(MkTransition)
+  , always
+  , runFSM
+  ) where
+
+import           Prelude (Show(..))
+import           Data.Bool
+import           Data.String
+import           Data.Maybe
+import           Data.Function
+import           Data.Monoid
+import           Control.Monad
+
+import           Test.QuickCheck.Gen
+import           Test.QuickCheck.Monadic
+import           Test.QuickCheck.Property
+
+-- | Defines a transition from state to state
+data Transition e s m = MkTransition {
+    -- | Display name (used in case of failure)
+    name :: String
+    -- | Predicate which determines of this transition is applicable for a given state 's'
+  , preCond :: s -> Bool
+    -- | Transition from state 's' given environment 'e'
+    --   Assertion logic can be placed in 'PropertyM m s'
+  , transition :: e -> s -> PropertyM m s
```

Comment:
Wondering whether we leave off the `e`? Consumers do this with `ReaderT` if they like.